src/persist-client/src/operators/shard_source.rs

```diff
@@ -224,20 +223,20 @@ where
     // This operator doesn't need to use a token because it naturally exits when its input
     // frontier reaches the empty antichain.
     builder.build(move |_caps| async move {
-        let Ok(mut lease_returner) = rx.await else {
+        let Ok(leases) = rx.await else {
             // Either we're not the chosen worker or the dataflow was shutdown before the
             // subscriber was even created.
             return;
         };
         while let Some(event) = completed_fetches.next().await {
-            let Event::Data(_cap, data) = event else {
+            let Event::Progress(frontier) = event else {
                 continue;
             };
-            for part in data {
-                lease_returner.return_leased_part(
-                    lease_returner.leased_part_from_exchangeable::<G::Timestamp>(part),
-                );
-            }
+            // If the frontier is not <= a particular time, that indicates that upstream is
+            // finished with that time.
+            leases
+                .borrow_mut()
+                .retain(|(time, _lease)| frontier.less_equal(time));
```

Comment:
Mild preference for pulling this out into a struct with some basic unit tests.

Also, calling retain every time the frontier advances could be `O(n^2)`, no? If so, let's fix that

Reply:
> Mild preference for pulling this out into a struct with some basic unit tests.

Aight, done.

> Also, calling retain every time the frontier advances could be O(n^2), no? If so, let's fix that

Yeah... somewhat mitigated by how the number of in-flight parts is only very large before the first frontier advancement, which clears the whole thing out.

But anyways I think we can just use a queue given our access pattern, so I'll switch to that!